modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java

```diff
@@ -7,102 +7,193 @@
+        final ValueHolder<MetadataResolvingException> exceptionValueHolder = new ValueHolder<>();
 
-    private void fillFieldValueMap(MetadataKey key, Class<?> type) throws MetadataResolvingException
-    {
-        Map<Integer, Field> metadataKeyPartMap = getAnnotatedFields(type, MetadataKeyPart.class)
-                .stream().collect(new ImmutableMapCollector<>(f -> f.getAnnotation(MetadataKeyPart.class).order(), identity()));
+        metadataType.accept(new MetadataTypeVisitor()
+        {
+            @Override
+            protected void defaultVisit(MetadataType metadataType)
+            {
+                exceptionValueHolder.set(buildException(String.format("'%s' type is invalid for MetadataKeyId parameters, use String type instead. Affecting component: '%s'", metadataKeyType.getSimpleName(), componentModel.getName())));
+            }
 
-        int order = 1;
-        while (!key.getChilds().isEmpty())
+            @Override
+            public void visitString(StringType stringType)
+            {
+                keyValueHolder.set(key.getId());
+            }
+
+            @Override
+            public void visitObject(ObjectType objectType)
+            {
+                try
+                {
+                    keyValueHolder.set(resolveMultiLevelKey(componentModel, key, metadataKeyType));
+                }
+                catch (MetadataResolvingException e)
+                {
+                    exceptionValueHolder.set(e);
+                }
+            }
+        });
+
+        if (exceptionValueHolder.get() != null)
         {
-            fieldValueMap.put(metadataKeyPartMap.get(order), key.getId());
-            key = key.getChilds().iterator().next();
-            order++;
+            throw exceptionValueHolder.get();
         }
-        fieldValueMap.put(metadataKeyPartMap.get(order), key.getId());
+
+        return keyValueHolder.get();
     }
 
+
     /**
-     * @return a new instance of the {@link MetadataKeyId} parameter {@code type} with the values of the
-     * passed {@link MetadataKey}.
-     * @throws MetadataResolvingException if the instantiation fails.
+     * Resolves the KeyIdObject for a MultiLevel {@link MetadataKeyId}
+     *
+     * @param componentModel model property of the {@link MetadataKeyId} parameter
+     * @param key            key containing the values of each level
+     * @return the KeyIdObject for the {@link MetadataKeyId} parameter
+     * @throws MetadataResolvingException
      */
-    public Object resolve() throws MetadataResolvingException
+    private static Object resolveMultiLevelKey(ComponentModel componentModel, MetadataKey key, Class metadataKeyType) throws MetadataResolvingException
     {
+        final Map<Field, String> fieldValueMap = fillFieldValueMap(metadataKeyType, key);
+
         Object metadataKeyId;
         try
         {
-            metadataKeyId = getType(type).newInstance();
+            metadataKeyId = metadataKeyType.newInstance();
```

Comment:
use ClassUtils whenever you need to create a new instance of a given class object